drivers/usb/device/usb_dc_nrfx.c

```diff
@@ -523,62 +526,33 @@ void usb_dc_nrfx_power_event_callback(nrf_power_event_t event)
 	}
 }
 
-/**
- * @brief Enable/Disable the HF clock
- *
- * Toggle the HF clock. It needs to be enabled for USBD data exchange
- *
- * @param on		Set true to enable the HF clock, false to disable.
- * @param blocking	Set true to block wait till HF clock stabilizes.
- *
- * @return 0 on success, error number otherwise
- */
-static int hf_clock_enable(bool on, bool blocking)
+/* Clock ready callback */
+static void hfxo_rdy_cb(struct onoff_manager *mgr, struct onoff_client *cli,
+			uint32_t state, int res)
 {
-	int ret = -ENODEV;
-	struct device *clock;
-	static bool clock_requested;
+	/* do nothing */
+}
 
-	clock = device_get_binding(DT_LABEL(DT_INST(0, nordic_nrf_clock)));
-	if (!clock) {
-		LOG_ERR("NRF HF Clock device not found!");
-		return ret;
+/* Stopping HFXO, algorithm supports case when stop comes before clock is
+ * started. In that case, it is stopped from the callback context.
+ */
+static int hfxo_stop(struct nrf_usbd_ctx *ctx)
+{
+	if (atomic_cas(&ctx->clk_single_req, 1, 0)) {
+		return onoff_cancel_or_release(ctx->hfxo_mgr, &ctx->hfxo_cli);
 	}
 
-	if (on) {
-		if (clock_requested) {
-			/* Do not request HFCLK multiple times. */
-			return 0;
-		}
-		ret = clock_control_on(clock, CLOCK_CONTROL_NRF_SUBSYS_HF);
-		while (blocking &&
-			clock_control_get_status(clock,
-						 CLOCK_CONTROL_NRF_SUBSYS_HF) !=
-					CLOCK_CONTROL_STATUS_ON) {
-		}
-	} else {
-		if (!clock_requested) {
-			/* Cancel the operation if clock has not
-			 * been requested by this driver before.
-			 */
-			return 0;
-		}
-		ret = clock_control_off(clock, CLOCK_CONTROL_NRF_SUBSYS_HF);
-	}
+	return 0;
+}
 
-	if (ret && (blocking || (ret != -EINPROGRESS))) {
-		LOG_ERR("HF clock %s fail: %d",
-			on ? "start" : "stop", ret);
-		return ret;
-	}
+static int hfxo_start(struct nrf_usbd_ctx *ctx)
+{
+	if (atomic_cas(&ctx->clk_single_req, 0, 1)) {
+		sys_notify_init_callback(&ctx->hfxo_cli.notify, hfxo_rdy_cb);
```

Comment:
Wouldn't `sys_notify_init_spinwait()` be sufficient here, is this empty callback needed for something?